core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java

```diff
@@ -366,32 +367,63 @@ public final void abort(Throwable cause) {
             return;
         }
 
-        final SubscriptionImpl newSubscription = new SubscriptionImpl(
-                this, AbortingSubscriber.get(cause), ImmediateEventExecutor.INSTANCE, EMPTY_OPTIONS);
-        downstreamSubscriptionUpdater.compareAndSet(this, null, newSubscription);
+        if (!subscribedToUpstreamUpdater.compareAndSet(this, 0, 1)) {
+            // Already subscribed to upstream so just abort upstream.
+            // upstream.abort(cause) might be called as well in delegate(StreamMessage<T> upstream) method
+            // which is perfectly fine.
+            final StreamMessage<T> upstream = this.upstream;
+            assert upstream != null;
+            upstream.abort(cause);
+            return;
+        }
 
+        // Abort upstream if it's set.
         final StreamMessage<T> upstream = this.upstream;
         if (upstream != null) {
             upstream.abort(cause);
+        }
+
+        final SubscriptionImpl newSubscription = new SubscriptionImpl(
+                this, AbortingSubscriber.get(cause), ImmediateEventExecutor.INSTANCE, EMPTY_OPTIONS);
```

Comment:
Q) Should we cache and reuse `newSubscription`? From my understanding, it do nothing but is a marker.

Reply:
`newSubscription` has `cause` and it's used for the late subscriber.
https://github.com/line/armeria/blob/9841500b5604df7450632fbdf08de20370c2a97a/core/src/main/java/com/linecorp/armeria/common/stream/CancellableStreamMessage.java#L110

Reply:
Understood. It would be possible to use `abortCause`, but it requires additional refactoring that is outside this PR's scope. Let's keep it as is.